### PR TITLE
Google Sign-in: Migrate the pop-up flow to the new Google Identity Services API

### LINKED ISSUE #64788: Google Sign-in: Migrate the pop-up flow to the new Google Identity Services API
### Details

Part of https://github.com/Automattic/wp-calypso/issues/64538 and blocked by https://github.com/Automattic/wp-calypso/issues/64787.

We want to adapt the Sign in with Google button so it works with the new Google Identity Services API.

The plan is not to introduce any changes, leaving the UX as it currently is. This ticket should handle the scenarios where a pop-up is presented to the user.

As finishing this task by itself does not complete the necessary work outlined in https://github.com/Automattic/wp-calypso/issues/64538, we need to put the changes behind a feature flag.

### Checklist

- [x] Consume the new Google Identity Services API in `components/social-buttons/google.js`
- [x] Pass the auth code that is given in the `initCodeClient` callback to the endpoint introduced in #64787
- [x] Receive and yield the tokens to the `responseHandler` callback
- [x] Treat the places where `responseHandler` is handled so it does not call deprecated Google API methods
- [x] Put all that under a feature flag

### PR DESCRIPTION
**Pull request by zaguiini**
### Proposed Changes

We are migrating to the new Google Identity Services API. This PR migrates the pop-up flow behind a feature flag so existing users won't be affected. The redirect flow will be handled by another PR.

### Testing Instructions

In these instructions, we will link a WPCOM account to a Google account. This is the easier path to test as the others involve setting multiple Google accounts and that's less than desirable.

- Create and activate a WPCOM account using an e-mail (https://10minutemail.net) and a password (do NOT Google login);
- Run `bin/calypso-secrets .config/secrets.php` in your WordPress.com sandbox;
- Copy the output of the script to `config/secrets.json` in your wp-calypso checkout;
- Build this branch with `NODE_ENV=production CALYPSO_ENV=production yarn run build`;
- Run these commands in the Calypso root folder so you can SSL the local version you just built:

```sh
openssl req \
    -newkey rsa:2048 \
    -x509 \
    -nodes \
    -keyout ./config/server/key.pem \
    -new \
    -out ./config/server/certificate.pem \
    -subj /CN=wordpress.com \
    -reqexts SAN \
    -extensions SAN \
    -config <(cat /System/Library/OpenSSL/openssl.cnf \
     <(printf '[SAN]\nsubjectAltName=DNS:wpcalypso.wordpress.com')) \
    -sha256 \
    -days 365
 
# Add it to the trusted certificates
sudo security add-trusted-cert -d -k $(security list-keychains | grep System | cut -d '"' -f 2 ) ./config/server/certificate.pem
```

- Kick off the server with `NODE_ENV=production HOST=wpcalypso.wordpress.com PORT=443 PROTOCOL=https CALYPSO_ENV=wpcalypso BUILD_TRANSLATION_CHUNKS=true node build/server.js`;
- Proxy `wpcalypso.wordpress.com` to `127.0.0.1`;
- Open the dev tools and paste `document.cookie='G_AUTH2_MIGRATION=informational'`.

Keep the DevTools open to assert a few details below...

#### Verify that the current flow still works

- Open https://wpcalypso.wordpress.com/me/security/social-login;
- Open the DevTools and check that the [deprecation notice](https://github.com/Automattic/wp-calypso/issues/64538) shows up in the console;
- Click on `Connect` and choose a Google account not yet linked to any WPCOM account.

It should connect the account successfully. You should also be able to see that `/me/social-login/connect` was called successfully in the network tab in the DevTools.

#### Verify that the new flow works

- Open https://wpcalypso.wordpress.com/me/security/social-login?flags=migration/sign-in-with-google;
- Click `Disconnect` to unlink the previously connected Google account;
- Open the DevTools;
- Check that the feature was activated by the `Config flag enabled via URL: migration/sign-in-with-google` message in the console;
- Check that the deprecation notice is not there anymore;
- Click on `Connect` and choose an account not yet linked to any WPCOM account.

It should connect the account successfully.  The network tab in the DevTools should show that both the calls to `/wp-login.php?action=exchange-social-auth-code` and `/me/social-login/connect` were issued successfully.

---

#### Verifying that social sign-ups work

The new flow should work all across the application since they're using the same component. If you'd like to try a sign-up instead of a social connection, here's how you can do so:

- Open https://wpcalypso.wordpress.com/start/user?flags=migration/sign-in-with-google;
- Open the DevTools;
- Check that the feature was activated by the `Config flag enabled via URL: migration/sign-in-with-google` message in the console;
- Check that the deprecation notice is not there anymore;
- Click on `Continue With Google` and choose an account not yet registered to WPCOM.

You should be redirected to the `/start/domains` page. In the background, requests to `/wp-login.php?action=exchange-social-auth-code` and `/users/social/new` should've been successfully issued.

Related to #64788.

### CODE REVIEW

File: client/components/social-buttons/google.js
Comment (zaguiini) [inline client/components/social-buttons/google.js:199]: We need to register this event.

File: config/development.json
Comment (zaguiini) [inline config/development.json:111]: I thought about keeping it disabled by default for all the environments until we have the redirect flow PR (which would include a commit enabling the feature flag for all envs except production). The migration is still incomplete even behind the feature flag, so no reason to enable this now since it's in a partially working state.

File: client/components/social-buttons/google.js
Comment (zaguiini) [inline client/components/social-buttons/google.js:39]: The https://www.googleapis.com/auth/userinfo.profile scope does not include the e-mail (in the new GSI APIs), which leads to failures since the account creation endpoint requires the e-mail for identification purposes.

**Review by zaguiini — COMMENTED** [review]
I did not see the consent screen when creating an account for a brand new Google account, which is good.

File: client/components/social-buttons/google.js
Comment (zaguiini) [reply client/components/social-buttons/google.js:199]: 924-gh-Automattic/tracks-events-registration. Will merge once approved by peers.

File: client/components/social-buttons/google.js
Comment (zaguiini) [inline client/components/social-buttons/google.js]: We cannot set `this.state.error` here because that'd disable the Google button and the user won't be able to try again. We need to dispatch a toast, global notification.

**Comment by vykes-mac** [comment]
I have tested the mentioned cases:

**Verify that the current flow still works**
✅ Check that the https://github.com/Automattic/wp-calypso/issues/64538 shows up in the console;
✅ Google account connect successfully
✅ You should also be able to see that /me/social-login/connect was called successfully in the network tab in the DevTools.

**Verify that the new flow works**
✅ Check that the feature was activated by the `Config flag enabled via URL: migration/sign-in-with-google` message in the console;
✅ Check that the deprecation notice is not there anymore;
✅ Google account connect successfully
✅ The network tab in the DevTools should show that both the calls to `/wp-login.php?action=exchange-social-auth-code` and `/me/social-login/connect` were issued successfully.

**Verifying that social sign-ups work**
✅ Check that the feature was activated by the `Config flag enabled via URL: migration/sign-in-with-google` message in the console;
✅ Check that the deprecation notice is not there anymore;
✅ Continue with google redirected to the `/start/domains` page. 
✅ In the background, requests to `/wp-login.php?action=exchange-social-auth-code` and `/users/social/new` should've been successfully issued

**Review by vykes-mac — APPROVED** [review]
code looks good and tests works as expected

File: client/components/social-buttons/google.js
Comment (vindl) [reply client/components/social-buttons/google.js:199]: All of the other similar events are prefixed with `calypso_login_social_login*`. Would it make sense to do that here too for consistency?

Comment (zaguiini) [reply client/components/social-buttons/google.js:199]: I'm not sure if "all" similar events have the prefix. For example, there's `calypso_login_social_button_failure`. 

However, I do think adding the "social" bit is a fair call. Something like `calypso_login_social_auth_code_exchange_failure` would be better indeed.

Comment (zaguiini) [reply client/components/social-buttons/google.js:199]: Resolved in https://github.com/Automattic/wp-calypso/pull/65549.

**Pull request merged**
